src/LDATaskS3.cpp

```diff
@@ -0,0 +1,385 @@
+          uint64_t size_temp;
+          char* mem_to_send = model->serialize_to_S3(size_temp);
+          upload_wih_bucket_id_fn(mem_to_send, size_temp,
+                                  upload_lock_indicators[0], cur_train_idx);
+
+          delete mem_to_send;
+
+          psint->send_time_dist(time_sample,
+                                time_download + time_update + time_get_model);
+
+          std::cout << "--------------------------\n";
+          std::cout << "Time to download from S3: " << time_download
+                    << std::endl;
+          std::cout << "Time to send update to server: " << time_update
+                    << std::endl;
+          std::cout << "Time to get model from server (whole): "
+                    << time_get_model << std::endl;
+          std::cout << "Time to create model: " << time_create_model
+                    << std::endl;
+          std::cout << "Time to sample: " << time_sample << std::endl;
+          std::cout << "--------------------------\n";
+          std::cout << "Avg Time to download from S3: " << time_download / count
+                    << std::endl;
+          std::cout << "Avg Time to send update to server: "
+                    << time_update / count << std::endl;
+          std::cout << "Avg Time to get model from server (whole): "
+                    << time_get_model / count << std::endl;
+          std::cout << "Avg Time to create model: " << time_create_model / count
+                    << std::endl;
+          std::cout << "Avg Time to sample: " << time_sample / count
+                    << std::endl;
+
+          std::cout << "--------------------------\n";
+          std::cout << "documents/sec: "
+                    << (double) total_sampled_docs /
+                           ((get_time_ms() - start_time) / 1000)
+                    << std::endl;
+          std::cout << "tokens/sec: "
+                    << (double) total_sampled_tokens /
+                           ((get_time_ms() - start_time) / 1000)
+                    << std::endl;
+
+          std::cout << "--------------------------\n";
+          if (full_iteration == 0) {
+            std::cout << "sec/iteration: N/A yet \n";
+          } else {
+            std::cout << "sec/iteration: "
+                      << (double) full_iteration /
+                             ((get_time_ms() - start_time) / 1000)
+                      << std::endl;
+          }
+
+          std::cout << "successfully exit\n";
+          break;
+        }
+      }
+
+      // if there's only one LDAStatistics assigned to this worker,
+      // simply reset (cur) to avoid uploading & downloading
+      if (end == start + 1) {
+        cur = 0;
+        continue;
+      }
+
+      // load the next LDAStatistics from S3
+      while (true) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        if (upload_lock_indicators[cur_train_idx - start] == -1) {
```

Comment:
When do we break out of this loop?

Reply:
Loop only breaks when the required number of iterations has been run [1]. 

[1] https://github.com/hyu596/cirrus-1/blob/lda_opt/src/LDATaskS3.cpp#L380